### ISSUE TITLE
Issue with CXformPushJoinBelowUnionAll Impacting Compilation Time

### ISSUE THREAD

**Issue by rntlqvnf** [greenplum-db/gpdb#17427]
## Feature Request

**Is your feature request related to a problem? Please describe:**

Hello,

I am a student developer currently working on GPORCA, specifically focusing on supporting complex queries that incorporate UNION ALL operations. We have encountered a significant issue where the CXformPushJoinBelowUnionAll transformation considerably increases compilation time when applied to queries involving multiple joins and UNION ALL operations.

For instance, in our tests:

The compilation time for SELECT * FROM A JOIN B JOIN C is 105ms.
The compilation time for SELECT * FROM (A1 UNION ALL A2) JOIN B JOIN (C1 UNION ALL C2) is 1414ms.

This slowdown appears to stem from how CXformPushJoinBelowUnionAll handles child expressions during the pushdown process. It creates new expressions for each child, except the first. 

<img width="1285" alt="스크린샷 2024-05-06 오전 11 10 43" src="https://github.com/greenplum-db/gpdb/assets/68107883/244d375f-5485-4e30-87ed-29d9709051f5">

This results in differing output column references among the copied expressions, causing subexpressions in these child expressions to be stored in separate groups within the memo table, leading to significant bloat.

**Describe the feature you'd like to request:**

Could you please take a look at this issue and suggest any possible solutions? Alternatively, is this behavior expected?

**Comment by THANATOSLAVA** [comment]
Yeah the current behavior is expected. In this approach, we make a deep copy of `other` as its columns have to be remapped. This PR walks through the implementation and the reasoning behind the GUC being off by default.

https://github.com/greenplum-db/gpdb/pull/15535

We can potentially look into making `other` a shared scan, but that itself would require thorough thinking.

**Comment by rntlqvnf** [comment]
Thank you for your kind reply!
